File: cpp/p021.cpp
#include <iostream>
#include <vector>

int compute(const int TARGET) {
	std::vector<int> amicable_pairs = {};
	int sum = 0;
	for (int num_a = 220; num_a <= TARGET; num_a++) {
		int div_a = 0;
		for (int i = 1; i < num_a; i++) {
			if (num_a % i == 0)
				div_a += i;
		}
		int div_b = 0;
		for (int i = 1; i < div_a; i++) {
			if (div_a % i == 0)
				div_b += i;
		}
		if (num_a == div_b && div_a != div_b) {
			bool duplicate = false;
			for (auto i : amicable_pairs) {
				if (i == div_a) {
					duplicate = true;
					break;
				}
			}
			if (!duplicate) {
				amicable_pairs.push_back(num_a);
				amicable_pairs.push_back(div_a);
				sum += num_a + div_a;
			}
		}
	}
	return sum;
}

int main() {
	std::cout << compute(10000) << std::endl;
	return 0;
}

File: python/p021.py
def compute(LIMIT):
	amicable_pairs = []
	sum = 0
	for num_a in range(220, LIMIT):
		div_a = 0
		for i in range(1, num_a):
			if num_a % i == 0:
				div_a += i
		div_b = 0
		for i in range(1, div_a):
			if div_a % i == 0:
				div_b += i
		if num_a == div_b and div_a != div_b:
			for i in amicable_pairs:
				if i == div_a:
					break
			else:
				amicable_pairs.append(div_a)
				amicable_pairs.append(div_b)
				sum += div_a + div_b
	return sum

if __name__ == "__main__":
	print(compute(10000))
